### ISSUE TITLE
Similar mAP when splitting data into train, val and test

### ISSUE THREAD

**Issue by n-patricia** [ultralytics/yolov5#13262]
### Question

Hi,
I use yolov5x with this setting (train: 70%, val: 10%, test: 20%)
```
train: images/train  # train images (relative to 'path') 128 images
val: images/val  # val images (relative to 'path') 128 images
test: images/test # test images (optional)
``` 
I get similar mAP starts from epoch 18. The columns here are precision, recall, mAP0.5, mAP0.95.
<img width="510" alt="Screen Shot 2024-08-15 at 05 20 30" src="https://github.com/user-attachments/assets/7339f0e8-66e0-4f69-ba7f-7a1809e315ba">

However when I don't divide the training data, the metrics keep increasing until 300 epochs.
```
train: images/  # train images (relative to 'path') 128 images
val: images/  # val images (relative to 'path') 128 images
test: # test images (optional)
```
<img width="520" alt="Screen Shot 2024-08-15 at 05 36 24" src="https://github.com/user-attachments/assets/a1b218e4-d30e-4fab-9afa-0e046da1176e">

I use hyp-scratch-low for both experiments. The reason I do this is bacause I want to try my custom dataset like GlobalWheat2020 after this. Do I miss anything?

**Comment by glenn-jocher** [comment]
@n-patricia hi there,

Thank you for sharing your observations and detailed results! It's great to see your interest in optimizing your training process with YOLOv5.

From your description, it seems like the issue might be related to the dataset split and the size of your dataset. Here are a few points to consider:

1. **Dataset Size and Split**: With only 128 images in your dataset, splitting it into train, val, and test sets might be causing each subset to be too small to effectively train and validate the model. Typically, we recommend having at least 1500 images per class for training to achieve good results. If possible, try to increase the size of your dataset.

2. **Training with Default Settings**: It's always a good idea to start with default settings to establish a performance baseline. Ensure that your dataset is well-labeled and representative of the deployment environment. You can refer to our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips//) for more detailed guidance.

3. **Validation and Test Sets**: The purpose of the validation set is to tune the model during training, while the test set is used to evaluate the final model performance. If you combine all images into the training set, the model might overfit, leading to artificially high metrics. Ensure that your validation set is representative and sufficiently large to provide meaningful feedback during training.

4. **Hyperparameters and Overfitting**: Using the `hyp.scratch-low.yaml` is a good start. However, if you notice overfitting or plateauing metrics, you might want to experiment with different hyperparameters or use techniques like data augmentation to improve generalization.

Here is an example of how you might structure your dataset configuration for a larger dataset:

```yaml
train: images/train  # train images (relative to 'path')
val: images/val  # val images (relative to 'path')
test: images/test # test images (optional)
```

And for training, you can use:

```bash
python train.py --data custom.yaml --weights yolov5x.pt --epochs 300
```

If you continue to face issues, please ensure you are using the latest version of YOLOv5 and provide additional details such as training logs, loss curves, and any other relevant metrics. This will help us diagnose the problem more effectively.

Feel free to reach out if you have any more questions or need further assistance. Happy training! 🚀

**Comment by n-patricia** [comment]
@glenn-jocher thank you for your response. I understand your point. The problem from this data is imbalanced dataset. I tried to select 40 classes, with range from 13k to 200 samples per class. I just randomized the images for train and val set, because it is difficult to select balance classes for each set. Do you think it is better to let the algorithm select the samples by itself because of the imbalanced problem?

**Comment by glenn-jocher** [comment]
Hi @n-patricia,

Thank you for your detailed follow-up! Imbalanced datasets can indeed pose challenges during training, but there are several strategies you can employ to mitigate these issues and improve your model's performance.

### Strategies for Handling Imbalanced Datasets

1. **Class Weights**: One effective approach is to use class weights to give more importance to underrepresented classes. YOLOv5 supports this through the `--weights` argument. You can manually set class weights in your training script or use a method to calculate them based on the frequency of each class.

2. **Data Augmentation**: Augmenting your dataset can help balance the classes by artificially increasing the number of samples for underrepresented classes. Techniques like rotation, flipping, and color jittering can be beneficial.

3. **Oversampling and Undersampling**: You can oversample the minority classes or undersample the majority classes to balance the dataset. This can be done manually or using libraries like `imbalanced-learn`.

4. **Focal Loss**: Focal Loss is designed to address class imbalance by down-weighting easy examples and focusing more on hard, misclassified examples. While YOLOv5 does not natively support Focal Loss, you can modify the loss function in the codebase if you're comfortable with custom implementations.

### Example: Using Class Weights

Here's an example of how you might set class weights in your training script:

```python
# Example of setting class weights
class_weights = [1.0, 2.0, 0.5, ...]  # Adjust these values based on your class distribution
```

### Automated Sample Selection

Regarding your question about letting the algorithm select the samples, YOLOv5 does not currently have an automated sample selection feature based on class balance. However, you can implement a custom data loader that ensures balanced batches during training.

### Next Steps

1. **Verify Latest Version**: Ensure you are using the latest version of YOLOv5 to benefit from the latest features and bug fixes.
2. **Experiment with Class Weights**: Try setting class weights to see if it improves your model's performance.
3. **Augment Your Data**: Use data augmentation techniques to increase the diversity and balance of your dataset.

Feel free to share your results or any further questions you might have. The YOLO community and Ultralytics team are here to help you achieve the best possible results with your model. Good luck! 🚀

**Comment by n-patricia** [comment]
Hi @glenn-jocher 

I've read about class weights from here 
https://github.com/ultralytics/yolov5/discussions/5604

So one of the approaches is to simply adjust weight in utils/loss.py
[https://github.com/ultralytics/yolov5/blob/7ebb5e5da673350d4c168cf60d01986a5e0f00cc/utils/loss.py#L91-L101](url)
